### PR TITLE
htlc auditing extension

### PR DESCRIPTION
**Pull request by adecaro**
This PR introduces the following capabilities:
- Allow an auditor to check if an input or an output contains an HTLC script.
- If an input or output contains an HTLC script, the Auditor should be able to extract the script and learn about the conditions.

As a bonus, this PR also improves on the validation of the HTLC script.

Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

File: integration/token/interop/views/accept.go [hunk 1]
```diff
@@ -19,13 +22,29 @@ func (a *AcceptCashView) Call(context view.Context) (interface{}, error) {
...
+	outputs, err := tx.Outputs()
+	assert.NoError(err, "failed to get transaction's outputs")
+	now := time.Now()
+	for i := 0; i < outputs.Count(); i++ {
+		output, err := htlc.ToOutput(outputs.At(i))
+		assert.NoError(err, "cannot get htlc output wrapper")
+		if !output.IsHTLC() {
+			continue
+		}
+		// check script details, for example make sure the deadline has not passed
```
Comment (HagarMeir) [inline integration/token/interop/views/accept.go:41]: `AcceptCashView` is only used to accept when there is an issue (according to latest topology) and so this code is never used.

File: integration/token/interop/views/auditor.go [hunk 2]
```diff
@@ -59,8 +63,33 @@ func (a *AuditView) Call(context view.Context) (interface{}, error) {
...
+	now := time.Now()
+	for i := 0; i < outputs.Count(); i++ {
+		output, err := htlc.ToOutput(outputs.At(i))
+		assert.NoError(err, "cannot get htlc output wrapper")
+		if !output.IsHTLC() {
+			continue
 		}
+		// check script details, for example make sure the deadline has not passed
+		script, err := output.Script()
+		assert.NoError(err, "cannot get htlc script from input")
```
Comment (HagarMeir) [inline integration/token/interop/views/auditor.go:91]: from output

File: integration/token/interop/views/auditor.go [hunk 3]
```diff
@@ -59,8 +63,33 @@ func (a *AuditView) Call(context view.Context) (interface{}, error) {
...
+	}
+
+	now := time.Now()
+	for i := 0; i < outputs.Count(); i++ {
+		output, err := htlc.ToOutput(outputs.At(i))
+		assert.NoError(err, "cannot get htlc output wrapper")
+		if !output.IsHTLC() {
+			continue
 		}
+		// check script details, for example make sure the deadline has not passed
```
Comment (HagarMeir) [inline integration/token/interop/views/auditor.go:89]: `script.Validate` checks more than just the deadline

File: token/services/interop/htlc/script.go [hunk 4]
```diff
@@ -11,6 +11,8 @@ import (
 	"encoding/json"
 	"time"
 
+	"github.com/pkg/errors"
+
```
Comment (HagarMeir) [inline token/services/interop/htlc/script.go:15]: please remove empty line and run go imports

File: token/services/interop/htlc/audit.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	owner, err := identity.UnmarshallRawOwner(i.Owner)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal owner")
+	}
+	script := &Script{}
+	err = json.Unmarshal(owner.Identity, script)
+	if err != nil {
+		return nil, err
```
Comment (HagarMeir) [inline token/services/interop/htlc/audit.go:49]: maybe add a description to this error

File: token/services/interop/htlc/audit.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	owner, err := identity.UnmarshallRawOwner(o.Owner)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to unmarshal owner")
+	}
+	script := &Script{}
+	err = json.Unmarshal(owner.Identity, script)
+	if err != nil {
+		return nil, err
```
Comment (HagarMeir) [inline token/services/interop/htlc/audit.go:86]: maybe add a description to this error

File: token/services/interop/htlc/script.go [hunk 7]
```diff
@@ -33,6 +35,30 @@ type Script struct {
 	HashInfo  HashInfo
 }
 
+// Validate performs the following checks:
+// - The sender must be set
+// - The recipient must be set
+// - The deadline must be after the passed time reference
+// - HashInfo must be Available
+func (s Script) Validate(timeReference time.Time) error {
```
Comment (HagarMeir) [inline token/services/interop/htlc/script.go:43]: `func (s *Script)`

File: integration/token/interop/views/accept.go (hunk 1 above)
Comment (adecaro) [reply integration/token/interop/views/accept.go:41]: aha, good catch. It is during Lock. We have already stuff there. I will update.

**Review by HagarMeir — APPROVED** [review]

**Pull request merged**